### PR TITLE
Allow users to send emails from alias domains

### PR DESCRIPTION
**Pull request by alexander255**
*Description of the issue/feature this PR addresses*:

Currently, if I have a domain `bla.com` that is an alias of `blub.com` and John has user account as `john@blub.com`, then will not be able to send e-mails as `john@bla.com` because the system will not recognize / communicate to Postfix that `john@bla.com` is the same as `john@blub.com` and hence Postfix will reject this attempt.

*Proposed Solution*:

When calculating the allowed SASL usernames for an email address, also include aliases from aliased domains in the result of the query. (This will also work for the primary username, since that is added as “internal” alias to that table as well.)

*Current behavior before PR*:

John is not able to send emails as `john@bla.com`.

*Desired behavior after PR is merged*:

John is able to send emails as `john@bla.com`, but still not as `marry@bla.com`.

*Caveats*:

I only tested the proposed query with PostgreSQL.

### CODE REVIEW

**Comment by alexander255** [comment]
Letting a PR sit is the best way to have its quality naturally improve :wink:

I just discovered a big issue with the previous query: Since I used `INNER JOIN` instead of `LEFT JOIN` no results will ever be produced by the alias lookup if a domain does not have any aliases. Will fix this shortly.

File: modoboa/admin/postfix_maps.py [hunk 1]
```diff
@@ -116,10 +116,14 @@ class SenderLoginMap(object):
...
-        " INNER JOIN admin_domain dom ON dom.id=mb.domain_id"
-        " INNER JOIN admin_alias al ON alr.alias_id=al.id "
-        "WHERE al.enabled=1 AND al.address='%s')"
+        "INNER JOIN modoboa_admin_aliasrecipient alr ON alr.r_mailbox_id=mb.id "
+        "INNER JOIN admin_domain dom ON dom.id=mb.domain_id "
+        "INNER JOIN admin_alias al ON alr.alias_id=al.id "
+        "LEFT JOIN admin_domainalias adom ON adom.target_id=dom.id "
+        "WHERE al.enabled=1 AND ("
+        "  al.address='%s' OR ("
+        "    adom.name='@d' AND al.address='%u'||'@'||dom.name"
```
[the remark this answers is not in the document]
Comment (alexander255) [reply modoboa/admin/postfix_maps.py]: Took me forever to get what your actual question was: Yes, it should have been `'%d'` not `'@d'` of course. Also I fixed the MySQL-ism wrt concat, I didn't know MySQL was being special here again.

**Comment by alexander255** [comment]
Here's the explanation I wrote while trying to understand your question, I've checked the queries again and compared all three versions for equivalency and -isms and things should hopefully be OK now :slightly_smiling_face: 

Assuming we have:

  * Mailbox: `.id=10; .domain_id=30; .address='john';…`
  * Alias-Recipient: `.id=20; .r_mailbox_id=10; .alias_id=40; …`
  * Domain: `.id=30; .name='example.com'; …`
  * Domain: `.id=31; .name='example.org'; …`
  * Alias: `.id=40; .enabled=true; .address='dummy@example.org'; .domain_id=31; …`
  * Domain-Alias: `.id=50; .name='example.net'; .target_id=31; …`

User `john@example.com` should be allowed send messages using at least any of the following addresses:

  * `john@example.com` (due to internal alias for user, not shown above)
  * `dummy@example.org` (due to alias)
  * `dummy@example.net` (due to domain alias on domain used in alias)

The first two already work, but the last one doesn't since there will not be any alias entries generated for domain aliases (there won't ever be an entry with address `dummy@example.net` in the alias table in our example). This change fixes this by extending the query to also allow for all aliases whose domain ID is being aliased in the Domain-Alias table.

The extension of query part therefore says: Also allow users to send messages if the domain-name of the message matches a domain alias whose aliasing target domain name matches an alias with the message's username and that target domain. (Read it a couple of times and it may make sense…)

**Comment by tonioo** [comment]
@alexander255 Makes sense. I've read the query a couple of times to make sure I understand it well ;)

**Comment by tonioo** [comment]
@alexander255 There is actually a bug with this new version and I don't understand why... You can't send emails using an address allowed by a catchall alias anymore (it was working before). Do you have any idea?

**Comment by alexander255** [comment]
@tonioo: Sorry, for the delay. Didn't even know about domain aliases, cool feature!
I tried to reproduce this by creating a dummy alias for `@example.com` and comparing the database's results to each of the queries, but when `dummy@example.com` didn't produce any results for the original query it quickly dawned on me what the problem is. …
Domain aliases are only matched by the third possible lookup strategy of [`smtpd_sender_login_maps`](http://www.postfix.org/postconf.5.html#smtpd_sender_login_maps): First it will try *user@domain*, then it may try just *user* (not in our case) and finally it will try *@domain*, which is what the domain alias function uses – note that this last form **does not have a user part**. As you may have guessed the bold part leads to the core of the problem as the new query now uses the `%u` substitution and by looking [the documentation of the psql driver](http://www.postfix.org/pgsql_table.5.html) we find:

>        query  The SQL query template used to search the database, where %s  is
>               a  substitute for the address Postfix is trying to resolve, e.g.
>                   query = SELECT replacement FROM aliases WHERE mailbox = '%s'
>
>               This parameter supports the following '%' expansions:
> 
>               […]
> 
>               %u     When the input key is an address of the form user@domain,
>                      %u is replaced by  the  SQL  quoted  local  part  of  the
>                      address.   Otherwise, %u is replaced by the entire search
>                      string.  **If the localpart is empty,  the  query  is  sup-
>                      pressed and returns no results.** (emphasis added)

Well f***. :slightly_frowning_face: 

I'll try to think about something, now that we know what the problem is…

**Comment by alexander255** [comment]
This query (PostgeSQL) works, but its quite long:

```sql
SELECT result.email FROM (
  -- By including a query index (qidx) we establish a priority order between
  -- the different queryies where all by the result of the lowest query will
  -- be ignored: This way a user that would normally be allowed to send
  -- messages for '@example.com' will still receive an error if there is
  -- another user that owns the specific alias 'john@example.com'.
  (
    SELECT 0 as qidx, email
    FROM core_user
    WHERE email='%s' AND is_active
  ) UNION (
    SELECT 1 as idx, mb.address || '@' || dom.name
    FROM admin_mailbox mb 
    INNER JOIN admin_senderaddress sad ON sad.mailbox_id=mb.id 
    INNER JOIN admin_domain dom ON dom.id=mb.domain_id 
    WHERE sad.address='%s'
  ) UNION (
    SELECT 2 as qidx, mb.address || '@' || dom.name
    FROM admin_mailbox mb 
    INNER JOIN modoboa_admin_aliasrecipient alr ON alr.r_mailbox_id=mb.id
    INNER JOIN admin_domain dom ON dom.id=mb.domain_id
    INNER JOIN admin_alias al ON alr.alias_id=al.id
    LEFT JOIN admin_domainalias adoma ON adoma.target_id=al.domain_id
    LEFT JOIN admin_domain adom ON adom.id=adoma.target_id
    WHERE al.enabled AND (
      al.address='%s' OR (
        adoma.name='%d' AND al.address='%u'||'@'||adom.name
    ))
  ) UNION (
    SELECT 3 as qidx, mb.address || '@' || dom.name
    FROM admin_mailbox mb 
    INNER JOIN modoboa_admin_aliasrecipient alr ON alr.r_mailbox_id=mb.id
    INNER JOIN admin_domain dom ON dom.id=mb.domain_id
    INNER JOIN admin_alias al ON alr.alias_id=al.id
    LEFT JOIN admin_domainalias adoma ON adoma.target_id=al.domain_id
    LEFT JOIN admin_domain adom ON adom.id=adoma.target_id
    WHERE al.enabled AND (
      al.address='@'||'%d' OR (
        adoma.name='%d' AND al.address='@'||adom.name
    ))
  )
) AS result ORDER BY result.qidx ASC LIMIT 1;
```

What makes this so unwieldy is that I'm trying emulate that fact that Postfix will only try the `@example.com` query if `dummy@example.com` did not return a result. Otherwise a user A, that owns `@example.com`, would be able to send messages as `dummy@example.com` even if that alias is owned by user B instead.

A benefit of this is that it gets rid of the extra database query when the normal `user@domain` fails. Let me know what you think.

**Pull request merged**